daffodil-io/src/main/scala/edu/illinois/ncsa/daffodil/processors/charset/BitsCharset.scala

```diff
@@ -0,0 +1,243 @@
+   * coder has not been reset, it is assumed we are in the middle of decoding
+   * many characters, and so no mandatory alignment is needed. However, if the
+   * coder was reset, then it is assumed that we may be unaligned at the start
+   * of decoding characters, and so we must check if we are mandatory aligned.
+   */
+  def isMandatoryAlignmentNeeded(): Boolean
+  def malformedInputAction(): CodingErrorAction
+  def onMalformedInput(action: CodingErrorAction): BitsCharsetDecoder
+  def unmappableCharacterAction(): CodingErrorAction
+  def onUnmappableCharacter(action: CodingErrorAction): BitsCharsetDecoder
+  def decode(in: ByteBuffer, out: CharBuffer, endOfInput: Boolean): CoderResult
+  protected def decodeLoop(in: ByteBuffer, out: CharBuffer): CoderResult
+
+  final def decode(in: ByteBuffer): CharBuffer = {
+    var n = scala.math.ceil(in.remaining() * averageCharsPerByte()).toInt
+    var out = CharBuffer.allocate(n)
+
+    if ((n == 0) && (in.remaining() == 0)) out
+    else {
+      reset()
+      var break = false
+      while (!break) {
+        var cr =
+          if (in.hasRemaining())
+            decode(in, out, true)
+          else
+            CoderResult.UNDERFLOW
+        if (cr.isUnderflow())
+          cr = flush(out)
+        if (cr.isUnderflow())
+          break = true
+        else if (cr.isOverflow()) {
+          n = 2 * n + 1; // Ensure progress; n might be 0!
+          val o = CharBuffer.allocate(n)
+          out.flip()
+          o.put(out)
+          out = o
+        } else
+          cr.throwException()
+      }
+      out.flip()
+      out
+    }
+  }
+}
+
+abstract class BitsCharsetEncoder {
+  def bitsCharset: BitsCharset
+  def averageBytesPerChar(): Float
+  def maxBytesPerChar(): Float
+  def averageBitsPerChar(): Float
+  def maxBitsPerChar(): Float
+  def replacement(): Array[Byte]
+  def replaceWith(newReplacement: Array[Byte]): BitsCharsetEncoder
+  def flush(out: ByteBuffer): CoderResult
+  def reset(): BitsCharsetEncoder
+  /**
+   * Used to determine if the data input stream must be aligned (if not already)
+   * for this encoding. Based on whether the coder has been reset. If the
+   * coder has not been reset, it is assumed we are in the middle of encoding
+   * many characters, and so no mandatory alignment is needed. However, if the
+   * coder was reset, then it is assumed that we may be unaligned at the start
+   * of encoding characters, and so we must check if we are mandatory aligned.
+   */
+  def isMandatoryAlignmentNeeded(): Boolean
+  def malformedInputAction(): CodingErrorAction
+  def onMalformedInput(action: CodingErrorAction): BitsCharsetEncoder
+  def unmappableCharacterAction(): CodingErrorAction
+  def onUnmappableCharacter(action: CodingErrorAction): BitsCharsetEncoder
+  def encode(in: CharBuffer, out: ByteBuffer, endOfInput: Boolean): CoderResult
+  protected def encodeLoop(in: CharBuffer, out: ByteBuffer): CoderResult
+}
+
```

Comment:
Scaladoc badly needed. This implements our BitsCharset, but by delegating to a Java Charset.